R/audit-trail.R

```diff
@@ -0,0 +1,200 @@
+        request_body <- jsonlite::toJSON(request_body, auto_unbox = TRUE) |> as.character()
+      }
+      private$request_body <- request_body
+    },
+    set_response_body = function(response_body) {
+      checkmate::assert_false(
+        typeof(response_body) == "list"
+      )
+      private$response_body <- response_body
+    },
+    set_event_type = function(event_type) {
+      private$event_type <- event_type
+    },
+    set_study_id = function(study_id) {
+      private$study_id <- study_id
+    },
+    set_response_code = function(response_code) {
+      private$response_code <- response_code
+    },
+    validate_log = function() {
+      checkmate::assert(
+        !private$disabled
+      )
+      if (is.null(private$event_type)) {
+        if (private$response_code == 404) {
+          # "soft" validation failure for 404 errors
+          # it might be just invalid endpoint
+          # so we don't want to fail the request
+          return(FALSE)
+        } else {
+          stop("Event type not set for audit log. Please set the event type using `audit_log_event_type`")
+        }
+      }
+      return(TRUE)
+    },
+    persist = function() {
+      checkmate::assert(
+        !private$disabled
+      )
+      db_conn <- pool::localCheckout(db_connection_pool)
+      values <- list(
+        private$request_id,
+        private$event_type,
+        private$study_id,
+        private$endpoint_url,
+        private$request_method,
+        private$request_body,
+        private$response_code,
+        private$response_body
+      )
+
+      values <- purrr::map(values, \(x) ifelse(is.null(x), NA, x))
+
+      DBI::dbGetQuery(
+        db_conn,
+        "INSERT INTO audit_log (
+          request_id,
+          event_type,
+          study_id,
+          endpoint_url,
+          request_method,
+          request_body,
+          response_code,
+          response_body
+        )
+        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)",
+        values
+      )
+    }
+  ),
+  private = list(
+    disabled = FALSE,
+    request_id = NULL,
+    event_type = NULL,
+    study_id = NULL,
+    endpoint_url = NULL,
+    request_method = NULL,
+    response_code = NULL,
+    request_body = NULL,
+    response_body = NULL
+  )
+)
+
+
+#' Set up audit trail
+#'
+#' This function sets up an audit trail for a given process. It uses plumber's hooks to log
+#' information before routing (preroute) and after serializing the response (postserialize).
+#'
+#' This function modifies the plumber router in place and returns the updated router.
+#'
+#' @param pr A plumber router for which the audit trail is to be set up.
+#' @param endpoints A list of regex patterns for which the audit trail should be enabled.
+#' @return Returns the updated plumber router with the audit trail hooks.
+#' @examples
+#' pr <- plumber::plumb("your-api-definition.R") |>
+#'   setup_audit_trail()
+setup_audit_trail <- function(pr, endpoints = list()) {
+  checkmate::assert(
```

Comment:
there is a special function for [checking lists](https://www.rdocumentation.org/packages/checkmate/versions/2.3.1/topics/checkList), maybe you will like it (we use it already somewhere)

Reply:
done